pkg/customresourcestate/config.go

```diff
@@ -161,14 +164,90 @@ type ConfigDecoder interface {
 	Decode(v interface{}) (err error)
 }
 
+// generateResources generates the set of possible resources from the varying V and, or K of the original config.
+func generateResources(resources []Resource, opts *options.Options) []Resource {
```

Comment:
Thanks for working on this!!

Does this only run at startup? What about in cases where CRDs are created after KSM initialization?

Reply:
I've moved the logic to run whenever that happens. However, something I noticed while doing so was that the entire `customresourcestate` logic was written tightly coupled with the assumption that there will always be single value for GVKs (which is what we are trying to make dynamic here), which consequently resulted in far-reaching changes, stemming from [the API change to support multiple values](https://github.com/kubernetes/kube-state-metrics/pull/1851/files#diff-66c98afaffbc7811d709fbcf53e6f793a28331819fb442eb455d47ad14dcfddfR40-R44).

Reply:
Here's one of the ways to look at this problem.

When VKs are not fixed, there can be two (hypothetical) states of the input file.

<details>
<summary>Initial</summary>

```yaml
kind: CustomResourceStateMetrics
spec:
  resources:
    - groupVersionKind:
        group: "foo"
        version: "*"
        kind: "*"
      metrics:
        - name: "foo"
          each:
            type: Info
            info:
              labelsFromPath:
                name: [metadata, name]
```

</details>

<details>
<summary>Resolved</summary>

```yaml
kind: CustomResourceStateMetrics
spec:
  resources:
    - groupVersionKind:
        group: "foo"
        version: "v1alpha1"
        kind: "Bar"
      metrics:
        - name: "foo"
          each:
            type: Info
            info:
              labelsFromPath:
                name: [metadata, name]
    - groupVersionKind:
        group: "foo"
        version: "v1"
        kind: "Bar"
      metrics:
        - name: "foo"
          each:
            type: Info
            info:
              labelsFromPath:
                name: [metadata, name]
    - groupVersionKind:
        group: "foo"
        version: "v2"
        kind: "Baz"
      metrics:
        - name: "foo"
          each:
            type: Info
            info:
              labelsFromPath:
                name: [metadata, name]
```

</details>

Now, assuming we implement https://github.com/kubernetes/kube-state-metrics/issues/1738#issuecomment-1241802049, [KSM's hot-reloading](https://github.com/kubernetes/kube-state-metrics/pull/1827/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R88) can be extended to [not only be triggered by file changes](https://github.com/kubernetes/kube-state-metrics/pull/1827/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R84), but also if there's a version or kind change in the cluster, under the defined group. This would allow us to avoid rewriting the CRS implementation for this particular feature, and even though https://github.com/kubernetes/kube-state-metrics/pull/1851#discussion_r1014530303 seems to be a feasible approach as well, I don't think this will affect the maintainability or scalability of the code-base in the future.

My point being, it is possible to base this feature of off an existing feature request, as opposed to rewriting the code. However, I'm not sure which way ultimately seems to make more sense.

Reply:
Basically, in the above approach we isolate the resolution logic from KSM internals, and make it seem so as if the configuration supplied a set of GVKs (in place of a fixed G, and varying VKs).

This approach may or may not be the right one, just thinking out loud here.

Reply:
> I've moved the logic to run whenever that happens.

Where is this? I've been looking at the code but I don't see where this is called upon changes to discovery.

Reply:
Hello Han. In the second fixup commit, this was moved from [here](https://github.com/kubernetes/kube-state-metrics/pull/1851/commits/ed9d193db18b23a911842d3fc50d573f6f9df43b#diff-bf91d16f16c99a97ad09eb6587f27d8fc075a68825d396e035c8100f864a7582L161) to [here](https://github.com/kubernetes/kube-state-metrics/pull/1851/commits/ed9d193db18b23a911842d3fc50d573f6f9df43b#diff-66c98afaffbc7811d709fbcf53e6f793a28331819fb442eb455d47ad14dcfddfR69). I won't recommend reviewing just yet, though, as the overall implementation is not fully done (due to my concerns above).

Reply:
I only see `fillGVR` called in `CreateClient`, does `CreateClient` get called on discovery changes?

Reply:
Yes, it's implemented on the `customResourceMetrics` type, and called everytime that change happens. This type consists of several [values](https://github.com/kubernetes/kube-state-metrics/pull/1851/files#diff-66c98afaffbc7811d709fbcf53e6f793a28331819fb442eb455d47ad14dcfddfL40-L42) that we need to resolve the varying VKs.

Reply:
This approach looks good to me.